### PR TITLE
Coding Style: Put "leaf" methods at the bottom

### PR DESCRIPTION
**Pull request by andrei-epure-sonarsource**
See https://trello.com/c/GkCCaAHa/13-coding-style-common-code-at-the-bottom

### CODE REVIEW

File: docs/coding-style.md [hunk 1]
```diff
@@ -40,6 +50,10 @@ Local functions should always be placed at the end of a method.
     }
 ```
 
+### Separation
```
Comment (andrei-epure-sonarsource) [inline docs/coding-style.md:53]: I moved this separately because everything above is about order, while this is about spacing

**Review by mary-georgiou-sonarsource — APPROVED** [review]
LGTM

File: docs/coding-style.md [hunk 2]
```diff
@@ -27,7 +27,17 @@ Static fields and properties should be placed before instance ones.
 
 Static methods are preferred to be after instance methods.
 
-Individual members must be separated by empty line, except sequence of constants, fields and abstract members. These should not be separated by empty lines.
+Methods which are called by other methods in the same category should be placed below the callers.
```
Comment (zsolt-kolbay-sonarsource) [inline docs/coding-style.md:30]: The only issue I have here is the word category. It's not clear at first reading whether it refers to one of the groups from the list above (e.g. constants is a category) or it's more refined (e.g. private static constants is a category). 
`Methods which are called by other methods in the same category should be placed below the callers.`
Maybe accessibility is a better word here, although that doesn't include static/non-static. Not entirely sure about it.

**Review by zsolt-kolbay-sonarsource — APPROVED** [review]
LGTM
Just one word which doesn't quite rub me right.

File: docs/coding-style.md (hunk 2 above)
Comment (andrei-epure-sonarsource) [reply docs/coding-style.md:30]: Thanks. I'll open an improvement. https://github.com/SonarSource/sonar-dotnet/pull/6787/files

File: docs/coding-style.md [hunk 3]
```diff
@@ -27,7 +27,17 @@ Static fields and properties should be placed before instance ones.
...
+
+```csharp
+public int Bar() => 2;
+
+int CallerOne(int x) => Foo() + x;
+
+int CallerTwo(int x) => Foo() + x + Bar();
```
Comment (pavel-mikula-sonarsource) [inline docs/coding-style.md:37]: I find it difficult to assign a semantic meaning to `Foo` and `Bar` and keep track of that while reading "foo plus x plus bar".
I think `Foo` and `Bar` should be avoided unless the names have no semantic meaning and no effect at all, like in some test cases.

I propose to rename it like this:
```
int CallerTwo(int x) => Leaf() + x + Other();
```
or like this, in case "Bar" was meant to highlight the fact that it's public.
```
int CallerTwo(int x) => Leaf() + x + PublicMethod();
```

File: docs/coding-style.md [hunk 4]
```diff
@@ -40,6 +50,10 @@ Local functions should always be placed at the end of a method.
     }
 ```
 
+### Separation
+
+Individual members must be separated by empty line, except sequence of constants, fields and abstract members. These should not be separated by empty lines.
```
Comment (pavel-mikula-sonarsource) [inline docs/coding-style.md:55]: This change is out of scope of the original thing that was discussed on the card.

We use this also for block of single-line properties, as those are the same as fields. And adding empty lines here makes it harder to read - to see boundaries of property-region and to understand what the "state" of the object is.

Comment (pavel-mikula-sonarsource) [reply docs/coding-style.md:55]: Examples:

Easy to read:
https://github.com/SonarSource/sonar-dotnet/blob/01bfee73f97a72e18093db7acb876bd985dcc104/analyzers/src/SonarAnalyzer.Common/AnalysisContext/SonarCompilationStartAnalysisContext.cs

Hard to read - hard to find that it changed from a sequence of properties to constructors or methods:
https://github.com/SonarSource/sonar-scanner-msbuild/blob/master/src/SonarScanner.MSBuild.Shim/ProjectInfoAnalysisResult.cs

or

https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.CFG/Sonar/Blocks/Block.cs

File: docs/coding-style.md (hunk 3 above)
Comment (andrei-epure-sonarsource) [reply docs/coding-style.md:37]: fair point, I've simplified also to remove the `x`

File: docs/coding-style.md (hunk 4 above)
Comment (andrei-epure-sonarsource) [reply docs/coding-style.md:55]: >This change is out of scope of the original thing that was discussed on the card.

there's no content change, it's "Clean as you code" by organizing the same information

Comment (andrei-epure-sonarsource) [reply docs/coding-style.md:55]: I won't open a discussion with the bubble for adding a `### Separation` and moving content from one place to another in the coding style

I really don't understand your comment

Comment (pavel-mikula-sonarsource) [reply docs/coding-style.md:55]: Indeed, I overlooked the red line in github UI 😒

Comment (pavel-mikula-sonarsource) [reply docs/coding-style.md:55]: I don't mind the `### Separation` section.
